### PR TITLE
[Projects] Fix non-remote source workflow running [1.2.x]

### PR DESCRIPTION
**Pull request by yonishelach**
1. **Fixing** #2689

2. **Removed `run_object` from `_PipelineRunStatus` and updating the returned engine object in the `_RemoteRunner.run()` to the inner engine** - The run_object become redundant, was only used by the `_RemoteRunner` which now returns in his `run` method the status with the inner engine which makes the code cleaner. In this way we can simply remove `wait_for_completion` and `get_run_status` from the `remote` engine. 
**Inner-use only**.

3. **Changing the `project` param in `get_state` to the `project.name`** - It is implemented only in the `_KFPRunner` and uses only the project's name.

4. **Adding `source` parameter to `project.run`** - Mainly for the `_RemoteRunner` to allow for example, to load project from a specific commit without changing the source of the project. This required to add the `save` param to `load_and_run` function. In other engines (`local` and `kfp`) the `source` is to make sure the code in the jobs is up-to-date.

### CODE REVIEW

File: mlrun/__main__.py [hunk 1]
```diff
@@ -1019,6 +1019,7 @@ def project(
                 schedule=schedule,
                 timeout=timeout,
                 overwrite=overwrite_schedule,
+                source=url,
```
Comment (yaronha) [inline mlrun/__main__.py:1022]: i assume its redundant, already initialized in the `.load`

File: mlrun/projects/pipelines.py [hunk 2]
```diff
@@ -397,20 +397,17 @@ def enrich_function_object(
...
-    def __init__(
-        self, run_id, engine, project, workflow=None, state="", run_object=None
-    ):
+    def __init__(self, run_id, engine, project, workflow=None, state=""):
         self.run_id = run_id
         self.project = project
         self.workflow = workflow
         self._engine = engine
         self._state = state
-        self.run_object = run_object  # for _RemoteRunner
```
Comment (Tankilevitch) [inline mlrun/projects/pipelines.py:408]: you are breaking client backwards compatibility please explain why?

File: mlrun/projects/pipelines.py [hunk 3]
```diff
@@ -528,11 +523,15 @@ def run(
...
         namespace=None,
+        source=None,
     ) -> _PipelineRunStatus:
         pipeline_context.set(project, workflow_spec)
         workflow_handler = _PipelineRunner._get_handler(
             workflow_handler, workflow_spec, project, secrets
         )
+        if source and not workflow_spec.run_local:
+            logger.info(f"setting project source: {source}")
+            project.set_source(source, pull_at_runtime=True)
```
Comment (Tankilevitch) [inline mlrun/projects/pipelines.py:534]: so if i want to run a specific pipeline and pass it a source it becomes the project source and change the default state of the project regarding the `pull_at_runtime`, why? can't we just apply it on the pipeline?

File: mlrun/projects/pipelines.py [hunk 4]
```diff
@@ -554,9 +553,7 @@ def run(
         return _PipelineRunStatus(id, cls, project=project, workflow=workflow_spec)
 
     @staticmethod
-    def wait_for_completion(
-        run_id, project=None, timeout=None, expected_statuses=None, run_object=None
-    ):
+    def wait_for_completion(run_id, project=None, timeout=None, expected_statuses=None):
```
Comment (Tankilevitch) [inline mlrun/projects/pipelines.py:556]: I see that it wasn't used at all, but r we familiar with the history of that param? can it break something removing it?

File: mlrun/projects/pipelines.py [hunk 5]
```diff
@@ -573,7 +570,7 @@ def wait_for_completion(
 
     @staticmethod
     def get_state(run_id, project=None):
-        project_name = project.metadata.name if project else ""
+        project_name = project or ""
```
Comment (Tankilevitch) [inline mlrun/projects/pipelines.py:573]: seems like before the project param was of type project, and now its str, let's add type hinting, also... i see that this is a public method, meaning user can use it (which now we break backwards compatibility for the use) I would also keep the old behavior and log a warning about the previous type deprecation.

File: mlrun/projects/pipelines.py [hunk 6]
```diff
@@ -708,11 +707,22 @@ def run(
...
+        source=None,
     ) -> typing.Optional[_PipelineRunStatus]:
         workflow_name = name.split("-")[-1] if f"{project.name}-" in name else name
         runner_name = f"workflow-runner-{workflow_name}"
         run_id = None
 
+        save, project_source = (
+            (False, source) if source else (True, project.spec.source)
+        )
+
```
Comment (Tankilevitch) [inline mlrun/projects/pipelines.py:719]: this one liner requires an explanation why we wouldn't want to save  if source is provided

Comment (Tankilevitch) [reply mlrun/projects/pipelines.py:719]: also, it the `project_source` is the exact name? 🤔

File: mlrun/projects/project.py [hunk 7]
```diff
@@ -1967,6 +1968,7 @@ def run(
                           for using the pre-defined workflow's schedule, set `schedule=True`
         :param timeout:   timeout in seconds to wait for pipeline completion (used when watch=True)
         :param overwrite: replacing the schedule of the same workflow (under the same name) if exists with the new one.
+        :param source:    remote source to overwrite the actual `project.spec.source` (used when engine is remote).
```
Comment (Tankilevitch) [inline mlrun/projects/project.py:1971]: why we want to override the `project.spec.source` on run method?

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Looks good, minor comments

File: mlrun/projects/pipelines.py (hunk 3 above)
Comment (yonishelach) [reply mlrun/projects/pipelines.py:534]: The `pull_at_runtime` is an error, removing. See 4. in description.

File: mlrun/projects/pipelines.py (hunk 5 above)
Comment (yonishelach) [reply mlrun/projects/pipelines.py:573]: adding type hint. this is also for inner use. see 3. in the description

File: mlrun/projects/pipelines.py (hunk 6 above)
Comment (yonishelach) [reply mlrun/projects/pipelines.py:719]: will add an explanation, what name u suggest? `url`?

File: mlrun/projects/project.py (hunk 7 above)
Comment (yonishelach) [reply mlrun/projects/project.py:1971]: you right, its not exactly overwrite... will rephrase, see 4. in description

File: mlrun/projects/pipelines.py (hunk 4 above)
Comment (yonishelach) [reply mlrun/projects/pipelines.py:556]: see 2. in description

**Pull request closed without merging**